draft-ietf-taps-arch.md

```diff
@@ -216,35 +216,53 @@ Transport Services implementations are meant to be flexible at connection establ
+
+The following two examples show non-equivalent Protocol Stacks:
+
+ - If the application requires preservation of message boundaries, a Protocol Stack that runs UDP as the top-level interface to the application is not equivalent to a Protocol Stack that runs TCP as the top-level interface. A UDP stack would allow an application to read out message boundaries based on datagrams sent from the remote system, whereas TCP does not preserve message boundaries on its own, but needs a framing protocol on top to determine message boundaries.
+
+- If the application specifies that it requires reliable transmission of data, then a Protocol Stack using UDP without any reliability layer on top would not be allowed to replace a Protocol Stack using TCP.
 
-## Scope for API and Implementation Definitions
+The following example shows an equivalent Protocol Stack:
 
-The Transport Services API is envisioned as the abstract model for a family of APIs that share a common way to expose transport features and encourage flexibility. The abstract API definition {{!I-D.ietf-taps-interface}} describes this interface and how it can be exposed to application developers.
+- If the application does not require reliable transmission of data, then a Protocol Stack that adds reliability could be regarded as an equivalent Protocol Stack as long as providing this would not conflict with any other application-requested properties.
 
-Implementations that provide the Transport Services API {{?I-D.ietf-taps-impl}} will vary due to system-specific support and the needs of the deployment scenario. It is expected that all implementations of Transport Services will offer the entire mandatory API. All implementations are expected to offer an API that is sufficient to use the distilled minimal set of features offered by transport protocols {{?I-D.ietf-taps-minset}}, including API support for TCP and UDP transport. However, some features provided by this API will not be functional in certain implementations. For example, it is possible that some very constrained devices might not have a full TCP implementation beneath the API.
+To ensure that security protocols are not incorrectly swapped, Transport Services systems MUST only select Protocol Stacks when the transport security protocols within the stacks are identical. Transport Services systems MUST NOT automatically fall back from secure protocols to insecure protocols, or to weaker versions of secure protocols. A Transport Services system MAY allow applications to explicitly specify that fallback to a specific other version of a protocol is allowed, e.g., to allow fallback to TLS 1.2 if TLS 1.3 is not available.
```

Comment:
The first sentence lays out a very strict requirement.
The second is in line with the first, but it describes something broader nevertheless - i.e., it makes me think: "if you say that they must be *identical*, why tell me that falling back to a weaker version is disallowed?
The third sentence, then, is an outright contradiction to the first.

I'm not sure how to best solve this... maybe precede the third sentence with "As an exception, ...", and swap order: move the second sentence to the end.

Reply:
Agree with @mwelzl that there's a contradiction here.
In previous text (now to be deleted), we start out by saying that the Transpor Services system MUST offer security features as requested by the application. Maybe we want to say this here, too, and then continue with SHOULD be identical and MUST NOT race or automatically fall back?

Reply:
Fixed up. Please check!

Reply:
LGTM; thanks for doing this!